# Pakiety i moduły

### Wyszukiwanie lokalizacji interpretera

Korzystając z języków skryptowych (takich jak Python) niejednokrotnie będziemy musieli skorzystać z programu powłoki (terminal, konsola). W systemie Windows będzie to cmd, w systemie Linux - bash, w systemie macOS - zsh. Jeżeli interpreter Pythona jest poprawnie zainstalowany w systemie operacyjnym, możemy sprawdzić jego wersję korzystając z polecenia powłoki: python -V. Jeżeli polecenie python -V zwróci wersję Pythona, to znaczy, że ścieżka do programu interpretera jest dołączona do zmiennej środowiskowej Path. W takim wypadku możemy sprawdzić gdzie znajduje się instalacja interpretera z poziomu Pythona za pomocą instrukcji: 

<pre>import sys
locate_python = sys.exec_prefix
print(locate_python)</pre>

In [22]:
import sys
print(sys.exec_prefix)

/Users/szymonmackowiak/opt/anaconda3


Interpreter wskaże ścieżkę do aktualnie uruchomionej instancji Pythona (jeżeli interpreter został uruchomiony w środowisku izolowanym, polecenie sys.exec_prefix wskaże miejsce w którym znajduje się środowisko izolowane).

Instrukcje interpretera można wykonywać również z linii poleceń powłoki, korzystając ze składni:
- w Windows: python -c "import nazwa_modulu; instrukcje" 
- w macOS: python3 -c "import nazwa_modulu; instrukcje"

Warto pamiętać, że wykonując dowolne polecenie w powłoce systemu, możemy zapoznać się z jego składnią i opcjonalnymi argumentami korzystając z polecenia: komenda -h, lub --help. Wyświetlą one dokument "pomocy".

### Sprawdzanie ścieżki do aktualnego katalogu roboczego

Aby sprawdzić ścieżkę do aktualnego katalogu roboczego (Current Working Directory) możemy skorzystać z modułu os:

<pre>import os
cwd = os.getcwd()
print(cwd)</pre>

In [23]:
import sys
print(sys.exec_prefix)

import os
print(os.getcwd())

/Users/szymonmackowiak/GIT/Podstawy-programowania-w-Pythonie/NOS_grupa_3


Aby zmienić aktualną lokalizację katalogu roboczego, możemy skorzystać z polecenia z modułu os: os.chdir(sciezka). Należy pamiętać, że w przypadku pracy z systemem operacyjnym możemy natrafić okoliczności, w ramach których niemożliwa będzie zmiana ścieżki katalogu roboczego (katalog wskazywany przez ścieżkę nie istnieje, nie posiadamy uprawnień do pracy w katalogu wskazywanym przez ścieżkę, ścieżka wzkazuje na obiekt, który nie jest katalogiem). Obsługa wymienionych sytuacji jest możliwa dzięki zdefiniowanym wyjątkom Pythona:
- katalog pod wskazaną nie istnieje: FileNotFoundError
- obiekt pod wskazaną ścieżką nie jest katalogiem: NotADirectoryError
- brak uprawnień: PermissionError
Poniższy przykład prezentuje sposób zmiany lokalizacji aktualnego katalogu roboczego:

In [24]:
import os

print(os.getcwd())

/Users/szymonmackowiak/GIT/Podstawy-programowania-w-Pythonie/NOS_grupa_3


In [27]:
os.chdir(r'/Users/szymonmackowiak/GIT/Podstawy-programowania-w-Pythonie/NOS_grupa_3')

In [28]:
print(os.getcwd())

/Users/szymonmackowiak/GIT/Podstawy-programowania-w-Pythonie/NOS_grupa_3


<pre>import os

path = '/projekt/app'

try:
    os.chdir(path)
    print(f"Aktualny katalog roboczy: {os.getcwd()}")
except FileNotFoundError:
    print(f"Katalog: {path} nie istnieje")
except NotADirectoryError:
    print(f"{path} nie jest katalogiem")
except PermissionError:
    print(f"Nie masz uprawnień do zmian w {path}")</pre>

### Ścieżki w Pythonie

Przez ścieżkę rozumiemy łańcuch znaków, który reprezentuje lokalizację danego zasobu w drzewie katalogów systemu. W zależności od systemu operacyjnego (Windows, macOS, Linux) ścieżki te mają różną strukturę oraz korzystają z różnych symboli (np. do rozdzielania ścieżek w systemach macOS i Linux służy symbol "/", tymczasem w systemach Windows jest to "\"). Aby skrypty w Pythonie były niezależne od systemu, na którym zostaną uruchomione, korzystamy z biblioteki pathlib, a w szczególności z jej klasy Path. Jeżeli ścieżkę umieścimy w konstruktorze klasy Path() zostatnie utworzony obiekt ścieżki. Co ciekawe, do tak powstałego obiektu możemy dostawiać kolejne elementy ścieżki za pomocą ukośników i łańcuchów znaków, a dopóki skrajny lewy element instrukcji będzie obiektem typu ścieżki - wynik całej linii będzie obiektem typu ścieżki. 

Przykład:
<pre>
from pathlib import Path

print(Path('app'))
print(type(Path('app')))

print(Path('app')/'dir')
print(type(Path('app')/'dir'))

print(Path('app', 'dir'))
print(type(Path('app', 'dir')))
</pre>

Klasa Path umośliwia również na oczytanie lokalizacji bieżącego katalogu roboczego (cwd) oraz na jago zmianę. 

Przykład:
<pre>
from pathlib import Path
import os

print(Path.cwd())
os.chdir('/Users')
print(Path.cwd())

</pre>

### Instalacja i lokalizacja pakietów i modułów

Instalacja bibliotek - biblioteki instalujemy za pomocą managerów pakietów. Najczęściej za pomocą mamagerów:
- pip
- conda

Powyższe managery uruchamiamy z poziomu wiersza poleceń powłoki, za pomocą komend:

Windows:
- py -m pip install nazwa_pakietu 
- conda install nazwa_pakietu

macOS:
- pip install nazwa_pakietu 
- conda install nazwa_pakietu

Manager pakietów pip:
Maganer pip jest podstawowym instalatorem pakietów w Pythonie. Pozwala na instalację pakietów z repozytorium Python Package Index oraz z innych repozytoriów. Więcej na temat pip i PyPI (instalacja, szczegóły techniczne) można znaleźć pod adresem https://pypi.org/project/pip/. Dokumentację pip można znaleźć pod adresem: https://pip.pypa.io/en/stable/cli/pip_install/.

Ze szczegółami obsługi managera pip można zapoznać się pod adresem: https://pip.pypa.io/en/stable/cli/
W szczególności warto zwrócić uwagę na najczęściej wykorzystywane polecenia:
- wyświetlanie pomocy managera pip: <pre>pip -h</pre> lub <pre>pip --help</pre>
- wyświetlanie listy zainstalowanych pakietów: <pre>pip list</pre>
- instalacja pakietu: <pre>pip install nazwa_pakietu</pre>
- deinstalacja pakietu: <pre>pip uninstall nazwa_pakietu</pre>
- instalacja listy pakietów z pliku tekstowego: <pre>pip install -r lista_pakietow.txt</pre>
- aktualizacja managera pip: <pre>pip install --upgrade pip</pre>

Pakiet a moduł:
- moduł - w praktyce oznacza plik z rozszerzeniem .py, który zawiera/wystawia zmienne, funkcje, klasy itd.
- pakiet - folder z modułami

Aby znaleźć ścieżkę, w której znajduje się pakiet lub moduł można skorzystać z instrukcji z poziomu interpretera:
- dla pakietów: korzystając z pola nazwa_pakietu.__path__
- dla konkretnego modułu: korzystając z pola nazwa_pakietu__file__

lub z poziomu linii poleceń powłoki:
- pip show nazwa_pakietu

Moduły i pakiety, które doinstalowujemy za pomocą managera pip najczęściej są instalowane w katalogu site-packages.

Kiedy importujemy biblioteki, interpreter Pythona najpierw sprawdza, czy dany zasób znajduje się w katalogu projektu. Jeżeli nie, wówczas przeszukuje katalog bibliotek zainstalowanych w środowisku interpretera. W związku z tym tworzone przez nas moduły lub pakiety możemy umieszczać w:
- katalogu projektu
- katalogu lib/PythonXX.XX/site-packages

In [35]:
import biblioteka_globalna

In [36]:
biblioteka_globalna.v

340

### Zadanie
Utwórz moduł stale_fizyczne zawierający następujące stałe:
- pi = 3.1415
- c = $3 \cdot 10^8$
- vd = 340
- G = $6.67 \cdot 10^{-11}$

Następnie zapisz plik pod dwiema nazwami:
- stale_fizyczne_local.py, który umieścisz w aktualnym katalogu roboczym
- stale_fizyczne_global.py, który umieścisz w katalogu site-packages (najpierw trzeba znaleźć jego położenie)

Utwórz program, w którym zaimportujesz oba moduły i zaprezentujesz ich działanie.

### Zadanie 

Utwórz moduł zawierający następujące funkcje statystyczne:
- obliczanie średniej arytmentycznej z listy
- obliczanie odchylenia standardowego z listy
- obliczanie współczynników regresji liniowej z dwóch list
- obliczanie współczynnika korelacji Pearsona z dwóch list

Następnie zapisz plik pod dwiema nazwami:
- statystyka.py, który umieścisz w aktualnym katalogu roboczym
- statystyka_global.py, który umieścisz w katalogu site-packages (najpierw trzeba znaleźć jego położenie)

Utwórz program, w którym zaimportujesz oba moduły i zaprezentujesz ich działanie.

# Środowisko izolowane

Tworząc projekt w Pythonie możemy natknąć się na problem związany z wersjami bibliotek lib środowiska. Przykładowo - na jednym komputerze projekt działa, na drugim nie ze względu na różnice w wersjach bibliotek i środowiska. W związku z tym dobrym pomysłem jest dołączenie do skryptu środowiska, w którym projekt będzie działać poprawnie. Takie środowisko nazywamy środowiskiem izolowanym, a najpopularniejszymni rozwiązaniami dla Pythona są:
- Veritual Env (venv)
- Conda Create

### Virtual Env

MacOS:
1. instalacja: <pre>pip3 install virtualenv</pre>
2. Tworzenie środowiska: <pre>python3 -m venv nazwa_srodowiska</pre>
3. Aktywacja środowiska: <pre>source nazwa_srodowiska/bin/activate</pre>
4. Dezaktywacja środowiska: <pre>deactivate</pre>

Windows:
1. instalacja: <pre>pip -m install virtualenv</pre> lub <pre>pip install venv</pre>
2. Tworzenie środowiska: <pre>python -m venv nazwa_srodowiska</pre>
3. Aktywacja środowiska: <pre>.\nazwa_srodowiska\Scripts\activate</pre>
4. Dezaktywacja środowiska: <pre>deactivate</pre>

### Conda Create

Alternatywnym sposobem tworzenia środowiska izolowanego jest skorzystanie z polecenia conda create. Warunkiem koniecznym jest posiadanie managera conda.

Windows/macOS
1. Tworzenie środowiska: <pre>conda create -n nazwa_srodowiska python=wersja_pythona</pre>
2. Aktywacja środowiska: <pre>conda activate nazwa_srodowiska</pre>
3. Dezaktywacja środowiska: <pre>deactivate</pre>

### Zadanie

Utwórz katalog projektu, w którym założysz środowisko izolowane, a następie aktywuj je i zainstaluj w nim (za pomocą maganera pakietów pip) wybrane biblioteki (np. numpy, jupyter, matplotlib, pandas). Następnie utwórz skrypt Pythona i zademonstruj (za pomocą polecenia sys.exec_prefix) w jakim środowisku pracuje. 

# Zaawansowane edytory kodu

Tworzenie kodu programu nieuchronnie wiąże się z popełnianiem błędów. Z kolei - kiedy tworzymy większy projekt lub aplikację, w jej skład zaczyna wchodzić więcej niż jeden plik. Nasza aplikacja dodatkowo może odnosić się lub korzystać z zasobów komputera, na której jest uruchamiana i wymaga od nas częstej pracy w programie powłoki systemowej (bash, cmd, zsh). W takiej sytuacji coraz większe znaczenie przy wytwarzaniu oprogramowania ma środowisko programistyczne. Na obecnym etapie nauki najważniejszymi cechami/funkcjonalnościami środowiska są:
- uzupełnianie składni (wpisując nazwy instrukcji z przestrzeni nazw, po naciśnięciu klawisza tab uzupełniana jest wprowadzana instrukcja lub pojawia się lista dostępnych instrukcji)
- linter (aplikacja analizująca cały nasz kod i wskazująca potencjalne błędy)
- połączenie z powłoką systemową
- podgląd zasobów w katalogu projektu

Do takich środowisk należą:
- Spyder: https://www.spyder-ide.org
- PyCharm: https://www.jetbrains.com/pycharm/
- Visual Studio Code: https://code.visualstudio.com